File: sdk/unity/Packages/gg.gameflow.gameserver/Samples~/DedicatedServer/GameFlowServerBootstrap.cs
using System.Threading.Tasks;
using GameFlow;
using GameFlow.Unity;
using UnityEngine;

// Minimal headless server bootstrap. Attach to a GameObject in your dedicated-server scene,
// or call the same sequence from your own startup code. Off GameFlow it runs in local mode
// automatically, so the same build works on your machine with no configuration.
public sealed class GameFlowServerBootstrap : MonoBehaviour
{
    private GameFlowClient _gf;

    private async void Start()
    {
        // Create the runner first: it pumps SDK callbacks onto the main thread and
        // sends a clean shutdown when the process quits.
        var runner = GameFlowRunner.Create();
        _gf = new GameFlowClient(new GameFlowOptions
        {
            Logger = new UnityDebugLogger(),
            Dispatcher = runner.Dispatcher,
        });
        runner.Bind(_gf);

        try
        {
            await _gf.Start();   // connect (with retries) or fall back to local mode
            // ... begin listening on your netcode transport here ...
            await _gf.Ready();   // health reporting starts automatically
            Debug.Log("[gameflow] server is ready for players");
        }
        catch (GameFlowException e)
        {
            Debug.LogError($"[gameflow] startup failed: {e.Code} — {e.Message}");
        }
    }

    // Call when a player session joins.
    public async Task OnPlayerJoined(string sessionId)
    {
        try
        {
            await _gf.Players.Connect(sessionId);
        }
        catch (ServerFullException e)
        {
            Debug.LogWarning($"[gameflow] rejecting {sessionId}: server full (capacity {e.Capacity})");
        }
        catch (PlayerAlreadyConnectedException)
        {
            // a reconnect of a live session — nothing to do
        }
    }

    // Call when a player session leaves.
    public async Task OnPlayerLeft(string sessionId)
    {
        await _gf.Players.Disconnect(sessionId);
    }
}

File: sdk/unreal/examples/DedicatedServerSample/GameFlowServerBootstrap.cpp
#include "GameFlowServerBootstrap.h"
#include "GameFramework/PlayerController.h"
#include "GameFramework/PlayerState.h"

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

UGameFlowSubsystem* AGameFlowServerBootstrap::GetGameFlow() const
{
    UGameInstance* GI = GetGameInstance();
    if (!GI)
    {
        UE_LOG(LogTemp, Warning, TEXT("[gameflow] GameInstance not available"));
        return nullptr;
    }
    UGameFlowSubsystem* Sub = GI->GetSubsystem<UGameFlowSubsystem>();
    if (!Sub)
    {
        UE_LOG(LogTemp, Warning, TEXT("[gameflow] UGameFlowSubsystem not found — is the GameFlow plugin enabled?"));
    }
    return Sub;
}

// ---------------------------------------------------------------------------
// GameMode overrides
// ---------------------------------------------------------------------------

void AGameFlowServerBootstrap::BeginPlay()
{
    Super::BeginPlay();

    // Skip on clients — this logic only runs on the dedicated server.
    if (!HasAuthority())
    {
        return;
    }

    UGameFlowSubsystem* GF = GetGameFlow();
    if (!GF)
    {
        return;
    }

    // Step 1 — Connect to the GameFlow runtime (or fall back to local mode).
    //   In sidecar mode this probes the runtime with retries and exponential
    //   backoff until ConnectTimeoutMs elapses; in local mode it returns
    //   immediately. The result delegate fires on the game thread.
    GF->Start(FGfVoidResult::CreateWeakLambda(GF, [GF](const FGameFlowError& Error)
    {
        if (!Error.IsOk())
        {
            UE_LOG(LogTemp, Error, TEXT("[gameflow] Start failed: %s"), *Error.Message);
            return;
        }

        UE_LOG(LogTemp, Log, TEXT("[gameflow] Connected (mode=%s)"),
            GF->GetMode() == EGameFlowMode::Sidecar ? TEXT("Sidecar") : TEXT("Local"));

        // Step 2 — Mark the server ready so the platform can route players to it.
        //   In sidecar mode this also starts the automatic health heartbeat.
        GF->Ready(FGfVoidResult::CreateWeakLambda(GF, [](const FGameFlowError& ReadyErr)
        {
            if (!ReadyErr.IsOk())
            {
                UE_LOG(LogTemp, Error, TEXT("[gameflow] Ready failed: %s"), *ReadyErr.Message);
                return;
            }
            UE_LOG(LogTemp, Log, TEXT("[gameflow] Server ready — accepting players"));
        }));
    }));

    // Optional: bind a Blueprint-assignable event for health-degraded alerts.
    GF->OnHealthDegraded.AddDynamic(this, &AGameFlowServerBootstrap::OnHealthDegradedEvent);
}

void AGameFlowServerBootstrap::PostLogin(APlayerController* NewPlayer)
{
    Super::PostLogin(NewPlayer);

    if (!NewPlayer || !NewPlayer->PlayerState)
    {
        return;
    }

    // Stable per-connection tracking key. GetPlayerId() is engine-assigned and
    // unique within the match (no OnlineSubsystem dependency). In production,
    // substitute your own account/session id from the allocation payload.
    const FString SessionId = FString::Printf(TEXT("player-%d"), NewPlayer->PlayerState->GetPlayerId());

    UGameFlowSubsystem* GF = GetGameFlow();
    if (!GF)
    {
        return;
    }

    GF->ConnectPlayer(SessionId, FGfVoidResult::CreateLambda([SessionId](const FGameFlowError& Error)
    {
        if (!Error.IsOk())
        {
            // PLAYER_ALREADY_CONNECTED is benign (reconnect of a live session).
            if (Error.Code != EGameFlowErrorCode::PlayerAlreadyConnected)
            {
                UE_LOG(LogTemp, Warning, TEXT("[gameflow] ConnectPlayer %s failed: %s"),
                    *SessionId, *Error.Message);
            }
        }
    }));
}

void AGameFlowServerBootstrap::Logout(AController* Exiting)
{
    Super::Logout(Exiting);

    APlayerController* PC = Cast<APlayerController>(Exiting);
    if (!PC || !PC->PlayerState)
    {
        return;
    }

    const FString SessionId = FString::Printf(TEXT("player-%d"), PC->PlayerState->GetPlayerId());

    UGameFlowSubsystem* GF = GetGameFlow();
    if (!GF)
    {
        return;
    }

    // DisconnectPlayer is idempotent — safe to call even if ConnectPlayer
    // was never confirmed (e.g. the player dropped before Start() finished).
    GF->DisconnectPlayer(SessionId, FGfRemoveResult::CreateLambda(
        [SessionId](const FGameFlowError& Error, bool /*bFound*/, const FPlayerList& /*List*/)
    {
        if (!Error.IsOk())
        {
            UE_LOG(LogTemp, Warning, TEXT("[gameflow] DisconnectPlayer %s failed: %s"),
                *SessionId, *Error.Message);
        }
    }));
}

// Fired by the OnHealthDegraded event when repeated health pings fail.
// Log or alert here; the SDK keeps the match alive regardless.
void AGameFlowServerBootstrap::OnHealthDegradedEvent()
{
    UE_LOG(LogTemp, Warning, TEXT("[gameflow] Health degraded — check sidecar connectivity"));
}

// ---------------------------------------------------------------------------
// Shutdown
//
// UGameFlowSubsystem::Deinitialize() stops the health heartbeat, closes the
// watch stream, and sends POST /shutdown automatically when the GameInstance
// is torn down (map change or process exit) — no explicit Shutdown() call is
// needed from the GameMode.
// ---------------------------------------------------------------------------
